pkg/refactor/logengine/logger.go

```diff
@@ -0,0 +1,118 @@
+package logengine
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/direktiv/direktiv/pkg/flow/database/recipient"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+)
+
+type LogNotify interface {
+	NotifyLogs(recipientID uuid.UUID, recipientType recipient.RecipientType)
+}
+
+type Loggerw struct {
+	Sugar *zap.SugaredLogger
+	Store LogStore
+	Pub   LogNotify
+}
+
+func (logger *Loggerw) Log(tags map[string]interface{}, level string, msg string, a ...interface{}) error {
+	msg = fmt.Sprintf(msg, a...)
+
+	if len(tags) == 0 {
+		logger.Sugar.Infow(msg)
+	} else {
+		ar := make([]interface{}, len(tags)+len(tags))
+		i := 0
+		for k, v := range tags {
+			ar[i] = k
+			ar[i+1] = v
+			i += 2
+		}
+		switch level {
+		case "info":
+			logger.Sugar.Infow(msg, ar...)
+		case "debug":
+			logger.Sugar.Debugw(msg, ar...)
+		case "error":
+			logger.Sugar.Errorw(msg, ar...)
+		case "panic":
+			logger.Sugar.Panicw(msg, ar...)
+		default:
+			logger.Sugar.Debugw(msg, ar...) // this should never happen
+		}
+	}
+
+	tags["level"] = level
+	err := logger.Store.Append(context.Background(), time.Now(), msg, tags)
+	if err != nil {
+		return err
+	}
+	id, _ := uuid.Parse(fmt.Sprintf("%s", tags["sender"]))
+	logger.Pub.NotifyLogs(id, recipient.RecipientType(fmt.Sprintf("%s", tags["senderType"])))
+
+	return nil
+}
+
+type cachedSQLLogStore struct {
+	store        LogStore
+	logQueue     chan *logMessage
+	logWorkersWG sync.WaitGroup
+}
+
+type logMessage struct {
+	t      time.Time
+	msg    string
+	fields map[string]interface{}
+}
+
+func (cls *cachedSQLLogStore) logWorker() {
+	defer cls.logWorkersWG.Done()
+
+	for {
+		l, more := <-cls.logQueue
+		if !more {
+			return
+		}
+		_ = cls.store.Append(context.Background(), l.t, l.msg, l.fields)
+	}
+}
+
+var _ LogStore = &cachedSQLLogStore{}
+
+// wraps the logstore with a caching layer.
+// defer to func() to shutdown the logger.
+func NewCachedLogger(store LogStore) (LogStore, func()) {
+	cls := cachedSQLLogStore{store: store}
```

Comment:
Better to explicitly all struct fields here.

Reply:
i removed the cached logstore since its not possible to notify the subscribers when new logs where successfully written to the db, because outside of the cached logger no one know when a specific log is posted to the db

Reply:
also i found out that its not possible to keep the output of the sugar-logger consistent with the logstore. But this should be an issue for an other time